src/libs/auth.ts

```diff
@@ -109,59 +149,146 @@ const getSigninArgs = (includeBillingScope) => {
 };
 
 export const signIn = async (includeBillingScope = false): Promise<User> => {
-  const args = getSigninArgs(includeBillingScope);
-  const user: User = await getAuthInstance().signinPopup(args);
+  // we should handle if we get back null or false here (if loading the authTokenFails)
+  const user: User | null | boolean = await loadAuthToken(includeBillingScope, true);
+  if (user instanceof User) {
+    const sessionId = uuid();
+    const sessionStartTime: number = Date.now();
+    authStore.update((state) => ({
+      ...state,
+      hasGcpBillingScopeThroughB2C: includeBillingScope,
+      sessionId,
+      sessionStartTime,
+    }));
+    Ajax().Metrics.captureEvent(Events.user.login, {
+      sessionStartTime: Utils.makeCompleteDate(sessionStartTime),
+    });
+    return user;
+  }
+  throw new Error('Auth token failed to load when signing in');
+};
 
-  const sessionId = uuid();
-  const sessionStartTime: number = Date.now();
-  const userJWT: string = user.id_token!;
-  const decodedJWT: JwtPayload = jwtDecode<JwtPayload>(userJWT);
-  const authTokenCreatedAt: number = (decodedJWT as any).auth_time; // time in seconds when authorization token was created
-  const authTokenExpiresAt: number = user.expires_at!; // time in seconds when authorization token expires, as given by the oidc client
-  const jwtExpiresAt: number = (decodedJWT as any).exp; // time in seconds when the JWT expires, after which the JWT should not be read from
+export const loadAuthToken = async (includeBillingScope = false, popUp = false): Promise<User | null | boolean> => {
+  // here we should update the auth token id
+  // might want to take refresh token and encrypt it to compare to a saved encrypted token
+  // to see if they are the same => track refresh update
+  // track lifespan of current authToken
+  // need to send information about the last authToken
+  const oldAuthTokenMetadata = authStore.get().authTokenMetadata; // this can be null (first token), so cover that case
```

Comment:
Are these TODO comments or notes left over from development?

Reply:
Sort of half and half. I will clean these up, but ultimately, I wanted to figure out how best to determine if we are using a new refresh token vs the same. Currently with these changes, we are just comparing the refresh tokens against each other, but if that is not ideal (for security of storing the token itself) I thought we could maybe store the hash of the refresh token instead. From what I can tell, the refresh token cannot be decoded (perhaps it is encrypted for the oidc client's use only) so I am comparing the refresh tokens against each other to see if they are the same.